### PR TITLE
[FLINK-29950] Refactor ResultSet to an interface

### PR DESCRIPTION
**Pull request by yuzelin**
## What is the purpose of the change

Currently, the `ResultSet` only contains Schema and data for execution result, witch is not enough for Client to display necessary information. So we need to add more fields to improve 'ResultSet'.

We can refactor the ResultSet to an interface and hide the detail of implementation of ResultSet, such as how to Ser/De the data in ResultSet.

## Brief change log
  - Change `ResultSet` to an interface and add new necessary methods.
  - Implement `ResultSetImpl`.
  - Add tests for new methods.
  

## Verifying this change

This change added tests and can be verified as follows:

  - All the tests in `SqlGatewayServiceITCase`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no
## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

**Comment by yuzelin** [comment]
@flinkbot run azure

File: flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/result/ResultFetcher.java [hunk 1]
```diff
@@ -220,4 +223,111 @@ public synchronized ResultSet fetchResults(long token, int maxFetchSize) {
...
+                            jobID,
+                            resultKind);
+
+            if (rows != null) {
+                resultFetcher.resultStore = ResultStore.DUMMY_RESULT_STORE;
+                resultFetcher.bufferedResults.addAll(rows);
+            } else {
+                resultFetcher.resultStore =
+                        new ResultStore(rowsIterator, TABLE_RESULT_MAX_INITIAL_CAPACITY);
+            }
```
Comment (fsk119) [inline flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/result/ResultFetcher.java:329]: Why not introducing 
```
    public static ResultFetcher fromTableResult(
            OperationHandle operationHandle, TableResult tableResult, boolean isQuery) {}

    public static ResultFetcher fromResults(
            OperationHandle operationHandle,
            ResolvedSchema resolvedSchema,
            List<RowData> results) {}
```

File: flink-table/flink-sql-gateway/src/test/java/org/apache/flink/table/gateway/service/SqlGatewayServiceITCase.java [hunk 2]
```diff
@@ -456,6 +450,100 @@ public void testGetOperationSchemaUntilOperationIsReady() throws Exception {
                 task -> assertThat(task.get()).isEqualTo(getDefaultResultSet().getResultSchema()));
     }
 
+    // ---------------------------------------------------------------------------------------------
+    // Executing statement tests
+    // ---------------------------------------------------------------------------------------------
```
Comment (fsk119) [inline flink-table/flink-sql-gateway/src/test/java/org/apache/flink/table/gateway/service/SqlGatewayServiceITCase.java:455]: I am prone to move the tests to the `SqlGatewayServiceStatementITCase` side.

File: flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/result/ResultFetcher.java [hunk 3]
```diff
@@ -220,4 +223,111 @@ public synchronized ResultSet fetchResults(long token, int maxFetchSize) {
     public ResultStore getResultStore() {
         return resultStore;
     }
+
+    private ResultSet buildEosResultSet() {
```
Comment (fsk119) [inline flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/result/ResultFetcher.java:227]: nit: buildEosResultSet -> buildEOSResultSet

**Review by fsk119 — COMMENTED** [review]
Thanks for your contribution. I left some comments.

**Review by fsk119 — COMMENTED** [review]
Thanks for your update. I left some comments.

File: flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/result/ResultFetcher.java [hunk 4]
```diff
@@ -220,4 +283,26 @@ public synchronized ResultSet fetchResults(long token, int maxFetchSize) {
...
+    private ResultSet buildEOSResultSet() {
+        return ResultSetImpl.newBuilder()
+                .resultType(ResultSet.ResultType.EOS)
+                .nextToken(null)
+                .resolvedSchema(resultSchema)
+                .data(Collections.emptyList())
+                .build();
+    }
+
+    private ResultSet buildPayloadResultSet() {
```
[the remark this answers is not in the document]
Comment (yuzelin) [reply flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/result/ResultFetcher.java:296]: Maybe Builder is not necessary. Using constructor is OK.

**Comment by fsk119** [comment]
@yuzelin I add a commit to improve some codes. Could you take a look?

**Comment by yuzelin** [comment]
> @yuzelin I add a commit to improve some codes. Could you take a look?

LGTM.

**Comment by yuzelin** [comment]
@flinkbot run azure

**Pull request merged**